function/transform/special.go

```diff
@@ -157,3 +157,93 @@ var Alias = function.MetricFunction{
 		return list, nil
 	},
 }
+
+// Derivative is special because it needs to get one extra data point to the left
+// This transform estimates the "change per second" between the two samples (scaled consecutive difference)
+var Derivative = newDerivativeBasedTransform("derivative", derivative)
+
+func derivative(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
+	}
+	return result, nil
+}
+
+// Rate is special because it needs to get one extra data point to the left
+// This transform functions exactly like Derivative but bounds the result to be positive.
+// That is, it returns consecutive differences which are at least 0.
+var Rate = newDerivativeBasedTransform("rate", rate)
+
+func rate(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
```

Comment:
Oh.. you're actually changing the behavior of `rate`. Why? `rate` should just be `nonNegativeDerivative`.

I'm guessing this is to smooth out spikes from missing data? We should accomplish this using `nanKeepLast` or whatever we called it. This approximation totally breaks for counters anyways: `[1000000, nil, 1000030]` with 30s resolution becomes `[0, nil, 33334]`.

As a side note, if any component values are nil, the result should be 0. For derivative, we could make the result nil.